File: Saniya/ads/binomialheap.c
#include <stdio.h>
#include <stdlib.h>

// Node structure
typedef struct Node {
    int key, degree;
    struct Node *parent, *child, *sibling;
} Node;

// Create node
Node* createNode(int key) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    newNode->key = key;
    newNode->degree = 0;
    newNode->parent = newNode->child = newNode->sibling = NULL;
    return newNode;
}

// Merge root lists
Node* mergeHeaps(Node* h1, Node* h2) {
    if (!h1) return h2;
    if (!h2) return h1;

    Node *head = NULL, *tail = NULL;

    if (h1->degree <= h2->degree) {
        head = tail = h1;
        h1 = h1->sibling;
    } else {
        head = tail = h2;
        h2 = h2->sibling;
    }

    while (h1 && h2) {
        if (h1->degree <= h2->degree) {
            tail->sibling = h1;
            tail = h1;
            h1 = h1->sibling;
        } else {
            tail->sibling = h2;
            tail = h2;
            h2 = h2->sibling;
        }
    }

    tail->sibling = (h1) ? h1 : h2;
    return head;
}

// Link trees
void linkTrees(Node* y, Node* z) {
    y->parent = z;
    y->sibling = z->child;
    z->child = y;
    z->degree++;
}

// Union heaps
Node* unionHeaps(Node* h1, Node* h2) {
    Node* newHead = mergeHeaps(h1, h2);
    if (!newHead) return NULL;

    Node *prev = NULL, *curr = newHead, *next = curr->sibling;

    while (next != NULL) {
        if ((curr->degree != next->degree) ||
            (next->sibling && next->sibling->degree == curr->degree)) {
            prev = curr;
            curr = next;
        } else {
            if (curr->key <= next->key) {
                curr->sibling = next->sibling;
                linkTrees(next, curr);
            } else {
                if (prev == NULL)
                    newHead = next;
                else
                    prev->sibling = next;

                linkTrees(curr, next);
                curr = next;
            }
        }
        next = curr->sibling;
    }

    return newHead;
}

// Insert
Node* insert(Node* heap, int key) {
    Node* temp = createNode(key);
    return unionHeaps(heap, temp);
}

// Find minimum
Node* getMin(Node* heap) {
    if (!heap) return NULL;

    Node *y = NULL, *x = heap;
    int min = x->key;

    while (x != NULL) {
        if (x->key < min) {
            min = x->key;
            y = x;
        }
        x = x->sibling;
    }

    return (y == NULL) ? heap : y;
}

// Reverse list
Node* reverseList(Node* node) {
    Node *prev = NULL, *curr = node, *next;

    while (curr) {
        next = curr->sibling;
        curr->sibling = prev;
        curr->parent = NULL;
        prev = curr;
        curr = next;
    }
    return prev;
}

// Extract minimum
Node* extractMin(Node* heap) {
    if (!heap) return NULL;

    Node *minNode = heap, *minPrev = NULL;
    Node *curr = heap, *prev = NULL;

    int min = heap->key;

    while (curr != NULL) {
        if (curr->key < min) {
            min = curr->key;
            minNode = curr;
            minPrev = prev;
        }
        prev = curr;
        curr = curr->sibling;
    }

    if (minPrev)
        minPrev->sibling = minNode->sibling;
    else
        heap = minNode->sibling;

    Node* child = reverseList(minNode->child);
    heap = unionHeaps(heap, child);

    printf("Extracted Min: %d\n", minNode->key);
    free(minNode);

    return heap;
}

// Display
void display(Node* heap) {
    if (!heap) {
        printf("Heap is empty\n");
        return;
    }

    Node* temp = heap;
    while (temp) {
        printf("B%d: ", temp->degree);

        Node* curr = temp;
        // Simple traversal (not full tree print)
        while (curr) {
            printf("%d ", curr->key);
            curr = curr->child;
        }
        printf("\n");

        temp = temp->sibling;
    }
}

// Main
int main() {
    Node* heap = NULL;
    int choice, value;

    while (1) {
        printf("\n--- Binomial Heap Menu ---\n");
        printf("1. Insert\n2. Find Min\n3. Extract Min\n4. Display\n5. Exit\n");
        printf("Enter choice: ");
        scanf("%d", &choice);

        switch (choice) {
            case 1:
                printf("Enter value: ");
                scanf("%d", &value);
                heap = insert(heap, value);
                break;

            case 2: {
                Node* min = getMin(heap);
                if (min)
                    printf("Minimum: %d\n", min->key);
                else
                    printf("Heap is empty\n");
                break;
            }

            case 3:
                heap = extractMin(heap);
                break;

            case 4:
                display(heap);
                break;

            case 5:
                exit(0);

            default:
                printf("Invalid choice!\n");
        }
    }
}

File: Anush/anush sangeeth  s/binomialheap.py
class Node:
    def __init__(self, key):
        self.key = key
        self.degree = 0
        self.parent = None
        self.child = None
        self.sibling = None


class BinomialHeap:
    def __init__(self):
        self.head = None

    def merge_roots(self, h1, h2):
        if not h1:
            return h2
        if not h2:
            return h1

        if h1.degree <= h2.degree:
            head = h1
            h1 = h1.sibling
        else:
            head = h2
            h2 = h2.sibling

        tail = head

        while h1 and h2:
            if h1.degree <= h2.degree:
                tail.sibling = h1
                h1 = h1.sibling
            else:
                tail.sibling = h2
                h2 = h2.sibling
            tail = tail.sibling

        tail.sibling = h1 if h1 else h2
        return head

    def link(self, y, z):
        y.parent = z
        y.sibling = z.child
        z.child = y
        z.degree += 1

    def union(self, other):
        new_heap = BinomialHeap()
        new_heap.head = self.merge_roots(self.head, other.head)

        if not new_heap.head:
            return new_heap

        prev = None
        curr = new_heap.head
        next_node = curr.sibling

        while next_node:
            if (curr.degree != next_node.degree or
                (next_node.sibling and next_node.sibling.degree == curr.degree)):
                prev = curr
                curr = next_node
            else:
                if curr.key <= next_node.key:
                    curr.sibling = next_node.sibling
                    new_heap.link(next_node, curr)
                else:
                    if prev:
                        prev.sibling = next_node
                    else:
                        new_heap.head = next_node
                    new_heap.link(curr, next_node)
                    curr = next_node
            next_node = curr.sibling

        return new_heap

    def insert(self, key):
        temp = BinomialHeap()
        temp.head = Node(key)
        self.head = self.union(temp).head

    def get_min(self):
        if not self.head:
            return None

        y = None
        x = self.head
        min_key = float("inf")

        while x:
            if x.key < min_key:
                min_key = x.key
                y = x
            x = x.sibling

        return y

    def extract_min(self):
        if not self.head:
            return None

        prev_min = None
        min_node = self.head
        prev = None
        curr = self.head
        min_key = curr.key

        while curr:
            if curr.key < min_key:
                min_key = curr.key
                prev_min = prev
                min_node = curr
            prev = curr
            curr = curr.sibling

        if prev_min:
            prev_min.sibling = min_node.sibling
        else:
            self.head = min_node.sibling

        child = min_node.child
        prev = None

        while child:
            next_child = child.sibling
            child.sibling = prev
            child.parent = None
            prev = child
            child = next_child

        temp = BinomialHeap()
        temp.head = prev

        self.head = self.union(temp).head

        return min_node.key


if __name__ == "__main__":
    heap = BinomialHeap()
    values = [10, 3, 7, 1, 14, 8]

    for v in values:
        heap.insert(v)

    print(heap.extract_min())
    print(heap.extract_min())
